### PR TITLE
Clean up our makefile

### PR DESCRIPTION
**Pull request by benjaoming**
### Summary

A lot of inaccuracies and missing documentation has entered the Makefile, which should ideally be a good entrypoint to understanding the source tree.

### Reviewer guidance

n/a

### References

n/a

----

### Contributor Checklist

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] Contributor has fully tested the PR manually
- [x] Screenshots of any front-end changes are in the PR description
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')

### Reviewer Checklist

- [ ] Automated test coverage is satisfactory
- [ ] Reviewer has fully tested the PR manually
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependencies files were updated (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Link to diff of internal dependency change is included
- [ ] CHANGELOG.rst is updated for high-level changes
- [ ] Contributor is in AUTHORS.rst

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -119,10 +144,10 @@ makemessages: assets makedocsmessages
 compilemessages:
 	python -m kolibri manage compilemessages
 
-syncmessages: ensurecrowdinclient uploadmessages downloadmessages distributefrontendmessages
```
Comment (benjaoming) [inline Makefile:127]: `distributefrontendmessages` no longer exists

File: Makefile [hunk 2]
```diff
@@ -119,10 +144,10 @@ makemessages: assets makedocsmessages
...
+syncmessages: ensurecrowdinclient uploadmessages downloadmessages
 
 ensurecrowdinclient:
-	ls -l crowdin-cli.jar || wget https://storage.googleapis.com/le-downloads/crowdin-cli/crowdin-cli.jar # make sure we have the official crowdin cli client
```
Comment (benjaoming) [inline Makefile:130]: Cleans up confusing output

File: Makefile [hunk 3]
```diff
@@ -140,6 +165,3 @@ dockerenvbuild: writeversion
 dockerenvdist: writeversion
 	docker run --env-file ./env.list -v $$PWD/dist:/kolibridist learningequality/kolibri:$$(cat kolibri/VERSION)
 
-kolibripippex:
```
Comment (benjaoming) [inline Makefile]: This seemed unused, couldn't find any mentions in the codebase

File: Makefile [hunk 4]
```diff
@@ -75,32 +99,33 @@ docs-user: clean-docs
...
 	rm -r kolibri/dist/*.dist-info  # pip installs from PyPI will complain if we have more than one dist-info directory.
 
 writeversion:
 	python -c "import kolibri; print(kolibri.__version__)" > kolibri/VERSION
+	@echo ""
+	@echo -n "Current version is now "
+	@cat kolibri/VERSION
+	@echo ""
 
-setrequirements:
```
Comment (benjaoming) [inline Makefile]: Merged with `buildconfig`

File: Makefile [hunk 5]
```diff
@@ -52,7 +76,7 @@ lint:
 	flake8 kolibri
 
 test:
-	python setup.py test
```
Comment (benjaoming) [inline Makefile:55]: Also: Need to make `pytest` run

File: Makefile (hunk 3 above)
Comment (lyw07) [reply Makefile]: I originally put it there to get the kolibripip.pex to use in `install_cexts.py`. But since we have the kolibripip.pex in the repository already, I agree that removing it makes our `Makefile` more clean! 👍

File: Makefile [hunk 6]
```diff
@@ -75,32 +99,33 @@ docs-user: clean-docs
...
-	pip install -t kolibri/dist -r $(REQUIREMENTS_CEXT_NOARCH) --no-deps
+	pip install -t kolibri/dist -r "requirements.txt"
+	python install_cexts.py --file "requirements/cext.txt" # pip install c extensions
+	pip install -t kolibri/dist -r "requirements/cext_noarch.txt" --no-deps
 	rm -r kolibri/dist/*.dist-info  # pip installs from PyPI will complain if we have more than one dist-info directory.
 
 writeversion:
 	python -c "import kolibri; print(kolibri.__version__)" > kolibri/VERSION
+	@echo ""
+	@echo -n "Current version is now "
```
Comment (lyw07) [inline Makefile]: When doing `make writeversion`, the result I got is
```
-n Current version is now 
0.8.0.dev020180124211410-git
```

I assume what we want here is something like 
```
Current version is now 0.8.0.dev020180124211410-git
```
I guess having `-n` here is to not print the trailing newline character? I found that in [Makefile](https://www.gnu.org/software/make/manual/html_node/Options-Summary.html#Options-Summary), the option `-n` stands for "don't execute", which overwrites `echo`'s option `-n`. That's why we are getting `-n Current version is now` instead of `Current version is now`. One way I think might work is changing the line to `@echo "Current version is now \c". But there may be a better way to handle this :)

File: Makefile [hunk 7]
```diff
@@ -75,32 +99,33 @@ docs-user: clean-docs
 docs: docs-user docs-developer
 
 release:
-	ls -l dist/
-	read "\nDo you want to upload everything in dist/*?\n\n CTRL+C to exit."
+	@ls -l dist/
+	@read "\nDo you want to upload everything in dist/*?\n\n CTRL+C to exit."
```
Comment (lyw07) [inline Makefile]: When I did `make release`, it only showed something like
```
total 237264
-rw-r--r--  1 lingyiwang  staff  57507242 Jan 24 16:23 kolibri-0.8.0.dev020180124211410-git.tar.gz
-rw-r--r--  1 lingyiwang  staff  63968248 Jan 24 16:23 kolibri-0.8.0.dev020180124211410_git-py2.py3-none-any.whl
```
I think this is because in `read TEXT`, `TEXT` stands for a parameter and stores the user input, so "\nDo you want to upload everything in dist/*?\n\n CTRL+C to exit." is not prompted.
I tried with adding a prompt option `@read -p "\nDo you want to upload everything in dist/*?\n\n CTRL+C to exit."` and then got
```
total 237264
-rw-r--r--  1 lingyiwang  staff  57507242 Jan 24 16:23 kolibri-0.8.0.dev020180124211410-git.tar.gz
-rw-r--r--  1 lingyiwang  staff  63968248 Jan 24 16:23 kolibri-0.8.0.dev020180124211410_git-py2.py3-none-any.whl
\nDo you want to upload everything in dist/*?\n\n CTRL+C to exit.
```
But I still got `\n` in the prompt instead of the real newlines. I'm not sure how to resolve this part of the problem.

File: Makefile [hunk 8]
```diff
@@ -1,24 +1,48 @@
-REQUIREMENTS=requirements.txt
-REQUIREMENTS_CEXT=requirements/cext.txt
-REQUIREMENTS_CEXT_NOARCH=requirements/cext_noarch.txt
-
 .PHONY: help clean clean-pyc clean-build list test test-all coverage docs release sdist
 
 help:
```
Comment (lyw07) [inline Makefile:3]: this is so helpful! 🎉

File: Makefile (hunk 6 above)
Comment (benjaoming) [reply Makefile]: Oh yeah wow, I didn't think this much through.. seems there's both the problem that it's a Makefile and then that it's `sh` doing stuff after :)

Thanks for spotting!

File: Makefile (hunk 7 above)
Comment (benjaoming) [reply Makefile]: Oh this is also a problem, yes.. it's hard to get newlines in the `read` command with `sh`.. but easy enough to still use `echo` so will do that with an empty `read`

File: Makefile (hunk 6 above)
Comment (benjaoming) [reply Makefile]: Well actually this didn't happen on my system, but will still use a different approach, just to avoid spreading a bad pattern.

File: Makefile [hunk 9]
```diff
@@ -75,32 +99,32 @@ docs-user: clean-docs
...
+	python install_cexts.py --file "requirements/cext.txt" # pip install c extensions
+	pip install -t kolibri/dist -r "requirements/cext_noarch.txt" --no-deps
 	rm -r kolibri/dist/*.dist-info  # pip installs from PyPI will complain if we have more than one dist-info directory.
 
 writeversion:
 	python -c "import kolibri; print(kolibri.__version__)" > kolibri/VERSION
+	@echo ""
+	@echo "Current version is now `cat kolibri/VERSION`"
 
-setrequirements:
```
Comment (rtibbles) [inline Makefile:98]: Why is this removed?

Comment (benjaoming) [reply Makefile:98]: It's all in the `buildconfig` target now... sort of a normalization, but also a bug fix... see: #3129

Comment (benjaoming) [reply Makefile:98]: No wait, it's not a bug fix... just merging to Make targets that always got applied sequentially.

**Comment by benjaoming** [comment]
@lyw07 I split out `install_cext.py` from the `staticdeps` target so it will not run inside tox environments, where the requirements are already installed via `requirements/test.txt`

File: Makefile [hunk 10]
```diff
@@ -1,24 +1,49 @@
...
+	@echo ""
+	@echo "make <command>"
+	@echo ""
+	@echo "Building"
+	@echo "--------"
+	@echo ""
+	@echo "dist: create distributed source packages in dist/"
+	@echo "pex: builds a portable .pex file for each .whl in dist/"
+	@echo "assets: builds javascript assets"
+	@echo "staticdeps: downloads/updates all static deps bundled into the dist"
```
Comment (indirectlylit) [inline Makefile]: can we clarify: "all static deps" -> "all static Python dependencies" or similar?

File: Makefile [hunk 11]
```diff
@@ -1,24 +1,49 @@
...
+	@echo "clean-pyc: remove Python file artifacts"
+	@echo "clean-assets: removes javascript assets"
+	@echo "writeversion: updates the kolibri/VERSION file"
+	@echo "release: package and upload a release"
+	@echo "buildconfig: [unsupported] runs a special script for building a source package with special requirements.txt"
+	@echo ""
+	@echo "Development"
+	@echo "-----------"
+	@echo ""
+	@echo "lint: check style with flake8"
```
Comment (indirectlylit) [inline Makefile]: check Python style with flake8

File: Makefile [hunk 12]
```diff
@@ -1,24 +1,49 @@
...
+	@echo ""
+	@echo "dist: create distributed source packages in dist/"
+	@echo "pex: builds a portable .pex file for each .whl in dist/"
+	@echo "assets: builds javascript assets"
+	@echo "staticdeps: downloads/updates all static deps bundled into the dist"
+	@echo "staticdeps-cext: downloads/updates C extensions for all supported platforms"
+	@echo "clean: restores code tree to a clean state"
+	@echo "clean-build: remove build artifacts"
+	@echo "clean-pyc: remove Python file artifacts"
+	@echo "clean-assets: removes javascript assets"
```
Comment (indirectlylit) [inline Makefile]: removes JavaScript build assets

File: Makefile [hunk 13]
```diff
@@ -1,24 +1,49 @@
...
+	@echo "release: package and upload a release"
+	@echo "buildconfig: [unsupported] runs a special script for building a source package with special requirements.txt"
+	@echo ""
+	@echo "Development"
+	@echo "-----------"
+	@echo ""
+	@echo "lint: check style with flake8"
+	@echo "test: run tests quickly with the default Python"
+	@echo "test-all: run tests on every Python version with tox"
+	@echo "coverage: check code coverage quickly with the default Python"
```
Comment (indirectlylit) [inline Makefile]: for test-related commands, let's specify when they also run JS tests

File: Makefile [hunk 14]
```diff
@@ -1,24 +1,49 @@
...
+	@echo "docs: generate all documentation"
+	@echo "docs-user: generate just the user docs"
+	@echo "docs-developer: generate just developer + API docs"
```
Comment (indirectlylit) [inline Makefile]: generate just developer & API docs

**Comment by indirectlylit** [comment]
whoops looks like Travis is failing actually

File: Makefile (hunk 9 above)
Comment (rtibbles) [reply Makefile:98]: This line `python build_tools/customize_requirements.py` has been omitted and is not reinstated anywhere else as far as I can see? This will make our custom builds impossible.

Comment (benjaoming) [reply Makefile:98]: Ah yes, is reinstated now!

File: Makefile (hunk 13 above)
Comment (benjaoming) [reply Makefile]: Sure thing, I don't see any of the JavaScript testing stuff recorded in the Makefile.. guess frontenders prefer other ways? :)

**Comment by benjaoming** [comment]
> Still not sure how custom requirements are set now. Note, that this was separated from the build config step, because it has to happen before we generate the staticdeps.

This wasn't intentional in this PR, while I do think we need to clean this up :) Thanks for spotting!

This PR has been rebased and all comments addressed - will work out the Travis errors, allthough they seem strangely unrelated.

**Comment by benjaoming** [comment]
Yup, now working, I think the rebase caused it to work again, because it was built against a broken `develop`

**Comment by benjaoming** [comment]
@indirectlylit I have addressed @rtibbles 's comments, is there anything else?

**Comment by benjaoming** [comment]
Merging this now because other workflow changes with the translation stuff will affect the Makefile.
